src/Features/CSharp/Portable/UseExpressionBodyForLambda/UseExpressionBodyForLambdaCodeStyleProvider.cs

```diff
@@ -0,0 +1,434 @@
+            : base(CSharpCodeStyleOptions.PreferExpressionBodiedLambdaExpressions,
+                   IDEDiagnosticIds.UseExpressionBodyForLambdaExpressionsDiagnosticId,
+                   UseExpressionBodyTitle,
+                   UseExpressionBodyTitle)
+        {
+        }
+
+        #region analysis
+
+        protected override bool DiagnosticsForOpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override (ImmutableArray<SyntaxKind>, Action<SyntaxNodeAnalysisContext, CodeStyleOption<ExpressionBodyPreference>>) GetSyntaxNodeAction()
+            => (ImmutableArray.Create(SyntaxKind.SimpleLambdaExpression, SyntaxKind.ParenthesizedLambdaExpression),
+                AnalyzeSyntax);
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, CodeStyleOption<ExpressionBodyPreference> option)
+        {
+            var declaration = (LambdaExpressionSyntax)context.Node;
+            var diagnostic = AnalyzeSyntax(context.SemanticModel, option, declaration, context.CancellationToken);
+            if (diagnostic != null)
+            {
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private Diagnostic AnalyzeSyntax(
+            SemanticModel semanticModel, CodeStyleOption<ExpressionBodyPreference> option,
+            LambdaExpressionSyntax declaration, CancellationToken cancellationToken)
+        {
+            if (CanOfferUseExpressionBody(option.Value, declaration))
+            {
+                var location = GetDiagnosticLocation(declaration);
+
+                var additionalLocations = ImmutableArray.Create(declaration.GetLocation());
+                var properties = ImmutableDictionary<string, string>.Empty.Add(nameof(UseExpressionBody), "");
+                return DiagnosticHelper.Create(
+                    CreateDescriptorWithId(UseExpressionBodyTitle, UseExpressionBodyTitle),
+                    location, option.Notification.Severity, additionalLocations, properties);
+            }
+
+            if (CanOfferUseBlockBody(semanticModel, option.Value, declaration, cancellationToken))
+            {
+                // They have an expression body.  Create a diagnostic to convert it to a block
+                // if they don't want expression bodies for this member.  
+                var location = GetDiagnosticLocation(declaration);
+
+                var properties = ImmutableDictionary<string, string>.Empty;
+                var additionalLocations = ImmutableArray.Create(declaration.GetLocation());
+                return DiagnosticHelper.Create(
+                    CreateDescriptorWithId(UseBlockBodyTitle, UseBlockBodyTitle),
+                    location, option.Notification.Severity, additionalLocations, properties);
+            }
+
+            return null;
```

Comment:
Note the simplicity of this method compared to what it has to be over here: https://github.com/dotnet/roslyn/pull/28765/files#diff-79b0bc1373e80fce4c6740e4326dd50bR30

There's no need for complex and error prone code like:

```
+            // If the user likes expression bodies, then we offer expression bodies from the diagnostic analyzer.
+            // If the user does not like expression bodies then we offer expression bodies from the refactoring provider.
+            // If the analyzer is disabled completely, the refactoring is enabled in both directions.
+            if (userPrefersExpressionBodies == forAnalyzer || (!forAnalyzer && analyzerDisabled))
```

We only get into this codepath in the first place if the user actually has this option enabled in a way such that analyzer should run.